### DISCUSSION TITLE
Why isn't my query data returning as expect?

### DISCUSSION THREAD

**Discussion by DiceBreakers** [pocketbase/pocketbase#3487] [Q&A] [answered]
Alright, this is my first project beyond basic HTML, so I'm hoping my mistake is obvious to someone more experience than I am.

I'm building a conversation starter web app using Sveltekit and Pocketbase. So far, I've got it to where users (myself, lol) can log in and add prompts (just an input field for the prompt with a series of checkboxes to select which categories a prompt fits in, then it automatically links the author to the prompt and sets whether that author's prompts are public or private). This is all working as I hoped.

Now that I've been able to add a few of my prompts, I'm trying to build the logic to retrieve them and am coming up short. Here is an excerpt of the data returned for the "generatedPrompts" array. Every other is undefined, and when they aren't, the "prompt" is simply the number of the data set and the author is always "4".

`[...
    {
        "prompt": 3,
        "author": 4
    },
    {},
    {},
    {
        "prompt": 6,
        "author": 4
    },
    {},
    {
        "prompt": 8,
        "author": 4
    },
    {},
    {
        "prompt": 10,
        "author": 4
    },
    {},
    {
        "prompt": 12,
        "author": 4
    },
    {},
    {
        "prompt": 14,
        "author": 4
    },
    {},
    {
        "prompt": 16,
        "author": 4
    },
    {},
    {
        "prompt": 18,
        "author": 4
    },
    {},
    {
        "prompt": 20,
        "author": 4
    },
    {},
...]`

Here is my +page.server.js for the code querying pocketbase and combining the prompts.

```
import { error } from '@sveltejs/kit';

export const actions = {
  generate: async ({ request, locals }) => {
    const data = await request.formData();
    const selectedCategories = data.getAll('categories') ?? [];

    try {
      const generatedPrompts = [];

      // Iterate through selected categories
      for (const category of selectedCategories) {
        // Fetch up to 5 prompts from each category where postPublic is true
        const records = await locals.pb.collection('prompts').getFullList({
          filters: {
            categories: category,
            postPublic: true, // Filter by "postPublic" field
          },
          fields: 'prompt, author', // Select only these fields
          sort: '-created',
          limit: 5, // Maximum of 5 prompts per category
        });

        generatedPrompts.push(...records.map((record) => ({
          prompt: record.prompt,
          author: record.author,
        })));
      }

      return {
        records: generatedPrompts,
      };
    } catch (err) {
      console.error('Error: ', err);
      throw error(err.status, "The robots didn't like something about that...");
    }
  },
};

```

And last, the generate function code that tries to parse it:

```
	async function generate(event: Event) {
  event.preventDefault();

  const formData = new FormData();

  selectedCategories.subscribe((list: CatItem[]) => {
    list.forEach(catItem => {
      if (catItem.checked) {
        formData.append('categories', catItem.value);
      }
    });
  });

  try {
    const response = await fetch('/play?/generate', {
      method: 'POST',
      body: formData,
    });

    if (response.ok) {
      const data = await response.json();
      const promptData = JSON.parse(data.data);

      // Extract prompts starting from the third element
      generatedPrompts = promptData.slice(2).map(promptObj => ({
        prompt: promptObj.prompt,
        author: promptObj.author,
      }));

      // Display the first prompt when generated
      displayNextPrompt();
    } else {
      console.error('Failed to generate prompts.');
    }
  } catch (error) {
    console.error('An error occurred:', error);
  }
}
```

Since my generatedPrompts console log is nonsense, I feel like my problem lies in one of these bits of code, but I'm enough out of my comfort zone that I'm not sure how to spot the mistake.

+page.server.js
`
      generatedPrompts = promptData.slice(2).map(promptObj => ({
        prompt: promptObj.prompt,
        author: promptObj.author,
      }));
`

or game.svelte
`
        generatedPrompts.push(...records.map((record) => ({
          prompt: record.prompt,
          author: record.author,
        })));
`

**Comment by ganigeorgiev** [comment]
Questions like this are really hard to answer because I don't have the full picture of your use case and I don't understand what are you trying to do.

What is the error that is returned? You could also log `err.response` and/or check the failed request in the Admin UI > Logs.

What is `categories` in the above - single or multiple `relation` field? What is the `category` value - an id, name, or something else?

Also the filter query parameter name must be `filter`, not _filters_ and the value should be string in the same format as the API rules.

Also there is no such parameter "limit". From where did you get this field?

If you want only the first 5 items of a collection, you can do something like (_assuming that `categories` is a multiple `relation` field and `category` is an id_):
```js
const result = await locals.pb.collection('prompts').getList(1, 5, {
  fields:    'prompt, author', // Select only these fields
  filter:    `categories.id ?= "${category.replace(/"/g, "'")}" && postPublic = true`,
  sort:      '-created',
  skipTotal: 1, // optional optimization to ignore the default count query
});

const records = result.items;
```

**Reply by DiceBreakers** [reply]
Thank you for your response. No error is returned in the web console or in the pocketbase logs, but your response has given me a few things to look into and correct. I've attached a couple screenshots of added prompts and the collection settings to clarify my use case. The categories field allows multiple choices from an established list.

I used chatGPT to help generate some of my code so that's where the filter typo and erroneous "limit" came from. I will review your code, re-read the official docs, and modify my code accordingly.


![Pocketbase Sample](https://github.com/pocketbase/pocketbase/assets/141854239/aaaa8366-5aa3-49b0-91a2-e414589c86a3)

![CollectionSettings](https://github.com/pocketbase/pocketbase/assets/141854239/d7a56055-652b-49ad-9b7b-68fe7b5ca88f)


Thanks again, I really appreciate your help!

**Reply by ganigeorgiev** [reply]
Since `categories` is a multiple `select` field, in this case you can use the [`:each` modifier](https://pocketbase.io/docs/api-rules-and-filters/#each-modifier) to apply a condition for each entry:

```js
const result = await locals.pb.collection('prompts').getList(1, 5, {
  fields:    'prompt, author', // Select only these fields
  filter:    `categories:each ?= "${category.replace(/"/g, "'")}" && postPublic = true`,
  sort:      '-created',
  skipTotal: 1, // optional optimization to ignore the default count query
});

const records = result.items;
```

**Reply by DiceBreakers** [reply]
Thanks again for your help. I was able to make some progress, I had to mess around with the around with the categories operators before I got a working syntax. The webpage still isn't grabbing the data correctly, but I'm going to take this one step at a time.

Now when I load the results directly through the web API, a query result looks like this:


`{"items":[{"author":"7l2e3m1wx6loaiq","prompt":"What musical group or artist have you been listening to recently?"},{"author":"7l2e3m1wx6loaiq","prompt":"What is your favorite animal and why?"},{"author":"7l2e3m1wx6loaiq","prompt":"What happens next?"},{"author":"7l2e3m1wx6loaiq","prompt":"What do you like to do when you're feeling playful?"},{"author":"7l2e3m1wx6loaiq","prompt":"Who has seen you at your silliest?"},{"author":"7l2e3m1wx6loaiq","prompt":"What's the meanest rumor that's been spread about you?"},{"author":"7l2e3m1wx6loaiq","prompt":"What's the meanest rumor you remember spreading about someone else?"},{"author":"7l2e3m1wx6loaiq","prompt":"If you could prank call a billionaire, who would you choose and what would you say?"},{"author":"7l2e3m1wx6loaiq","prompt":"How have your political views changed as you've matured?"},{"author":"7l2e3m1wx6loaiq","prompt":"If you could have dinner with one historical figure, who would you choose?"}],"page":1,"perPage":10,"totalItems":-1,"totalPages":-1}`


That author string is my users id key, but I've been trying to use expand to link directly to the author.username field of that relational field.



```
    try {
      const generatedPrompts = [];

        // Convert the category value to a string and then replace double quotes
        const categoriesString = selectedCategories.toString().replace(/,/g, '"||categories~"');

        const result = await locals.pb.collection('prompts').getList( 1, 10, {
			filter: `(categories~"${categoriesString}") && postPublic = true`,
			fields: 'prompt,author',
			expand: 'author',
        	sort: '-created',
			skipTotal: 1,
        });

		const records = result.items;

        generatedPrompts.push(...records.map((record) => ({
          prompt: record.prompt,
          author: record.author.username,
        })));
      

      return {
        records: generatedPrompts,
      };
    } catch (err) {
      console.error('Error: ', err);
      throw error(err.status, "The robots didn't like something about that...");
    }
  },
};

```

I've tried a few variations, like expand: 'author.username', or the documentation seemed to indicate you could use expand inside of fields, so I tried fields: 'prompt, expand.author.username'. I couldn't get any of those variations to work though.


![authorRelation](https://github.com/pocketbase/pocketbase/assets/141854239/7dee6099-efb4-4146-8e0b-99c61877e2ee)



Edit- Hmm. I was hoping to solve this one problem on my own before going to bed, but this one got the better of me as well.

The pocketbase logs URL spits out mostly good data like what I sent above (just need to fix the author username field). But for some reason the data the browser receives doesn't look like that, it looks like this

```
{
    "type": "success",
    "status": 200,
    "data": "[{\"records\":1},[2,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43],{\"prompt\":3,\"author\":4},\"What musical group or artist have you been listening to recently?\",\"7l2e3m1wx6loaiq\",{\"prompt\":6,\"author\":4},\"What is your favorite animal and why?\",{\"prompt\":8,\"author\":4},\"What happens next?\",{\"prompt\":10,\"author\":4},\"What do you like to do when you're feeling playful?\",{\"prompt\":12,\"author\":4},\"Who has seen you at your silliest?\",{\"prompt\":14,\"author\":4},\"What's the meanest rumor that's been spread about you?\",{\"prompt\":16,\"author\":4},\"What's the meanest rumor you remember spreading about someone else?\",{\"prompt\":18,\"author\":4},\"If you could prank call a billionaire, who would you choose and what would you say?\",{\"prompt\":20,\"author\":4},\"How have your political views changed as you've matured?\",{\"prompt\":22,\"author\":4},\"If you could have dinner with one historical figure, who would you choose?\",{\"prompt\":24,\"author\":4},\"Should we live with \\\"no regrets\\\"? Why or why not?\",{\"prompt\":26,\"author\":4},\"What's a habit that you would like to change?\",{\"prompt\":28,\"author\":4},\"If you were stuck on a desert island with only one music album, what would you choose?\",{\"prompt\":30,\"author\":4},\"What was your favorite movie when you were a teenager?\",{\"prompt\":32,\"author\":4},\"What's your favorite place in nature?\",{\"prompt\":34,\"author\":4},\"Where feels the most like home for you?\",{\"prompt\":36,\"author\":4},\"Have you ever changed your mind about something big? If so, share about it.\",{\"prompt\":38,\"author\":4},\"What's a trait that you admire in others?\",{\"prompt\":40,\"author\":4},\"What's something silly you remember doing as a kid?\",{\"prompt\":42,\"author\":4},\"What do you wish you knew 20 years ago?\",{\"prompt\":44,\"author\":4},\"Share something you struggle with and would like to work on.\"]"
}
```


Did some more last minute searching before I sleep. I found an old post of yours where you mention expanding a relation field and needing view API rules. My users API rules right now are all simple, the view rule is as follows. I'm guessing I need to update this somehow for expand to work?

id = @request.auth.id

Thanks again!

**Comment by DiceBreakers** [comment]
Okay, made a bit more progress.

First step was the view rule I mentioned above, the query wasn't able to expand while that was in place. I cleared it blank and now this structure in the server query returns decent results. Is there a security issue with leaving the View API rule blank on the users collection?

```
		let records = await locals.pb.collection('prompts').getFullList({
      filter: `(categories~"${categoriesString}") && postPublic = true`,
      expand: 'author',
      fields: 'expand.author.username,prompt',
      sort: '-created',
			skipTotal: 1,
        });
```

Has the API return results with the human readable author username listed.

`{"items":[{"expand":{"author":{"username":"stumblios"}},"prompt":"What musical group or artist have you been listening to recently?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"What is your favorite animal and why?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"What happens next?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"What do you like to do when you're feeling playful?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"Who has seen you at your silliest?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"What's the meanest rumor that's been spread about you?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"What's the meanest rumor you remember spreading about someone else?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"If you could prank call a billionaire, who would you choose and what would you say?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"How have your political views changed as you've matured?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"If you could have dinner with one historical figure, who would you choose?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"Should we live with \"no regrets\"? Why or why not?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"What's a habit that you would like to change?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"If you were stuck on a desert island with only one music album, what would you choose?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"What was your favorite movie when you were a teenager?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"What's your favorite place in nature?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"Where feels the most like home for you?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"Have you ever changed your mind about something big? If so, share about it."},{"expand":{"author":{"username":"stumblios"}},"prompt":"What's a trait that you admire in others?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"What's something silly you remember doing as a kid?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"What do you wish you knew 20 years ago?"},{"expand":{"author":{"username":"stumblios"}},"prompt":"Share something you struggle with and would like to work on."}],"page":1,"perPage":200,"totalItems":-1,"totalPages":-1}`

Now the problem seems to just be with how this is data is returned by the webserver to the browser. Somewhere along the line, the data turns into this:

```
{
    "type": "success",
    "status": 200,
    "data": "[{\"records\":1},[2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42],{\"prompt\":3,\"username\":-1},\"What musical group or artist have you been listening to recently?\",{\"prompt\":5,\"username\":-1},\"What is your favorite animal and why?\",{\"prompt\":7,\"username\":-1},\"What happens next?\",{\"prompt\":9,\"username\":-1},\"What do you like to do when you're feeling playful?\",{\"prompt\":11,\"username\":-1},\"Who has seen you at your silliest?\",{\"prompt\":13,\"username\":-1},\"What's the meanest rumor that's been spread about you?\",{\"prompt\":15,\"username\":-1},\"What's the meanest rumor you remember spreading about someone else?\",{\"prompt\":17,\"username\":-1},\"If you could prank call a billionaire, who would you choose and what would you say?\",{\"prompt\":19,\"username\":-1},\"How have your political views changed as you've matured?\",{\"prompt\":21,\"username\":-1},\"If you could have dinner with one historical figure, who would you choose?\",{\"prompt\":23,\"username\":-1},\"Should we live with \\\"no regrets\\\"? Why or why not?\",{\"prompt\":25,\"username\":-1},\"What's a habit that you would like to change?\",{\"prompt\":27,\"username\":-1},\"If you were stuck on a desert island with only one music album, what would you choose?\",{\"prompt\":29,\"username\":-1},\"What was your favorite movie when you were a teenager?\",{\"prompt\":31,\"username\":-1},\"What's your favorite place in nature?\",{\"prompt\":33,\"username\":-1},\"Where feels the most like home for you?\",{\"prompt\":35,\"username\":-1},\"Have you ever changed your mind about something big? If so, share about it.\",{\"prompt\":37,\"username\":-1},\"What's a trait that you admire in others?\",{\"prompt\":39,\"username\":-1},\"What's something silly you remember doing as a kid?\",{\"prompt\":41,\"username\":-1},\"What do you wish you knew 20 years ago?\",{\"prompt\":43,\"username\":-1},\"Share something you struggle with and would like to work on.\"]"
}
```

I feel like the problem is most likely this code on the server so time to educate myself on how map is supposed to function.

```
		generatedPrompts = records.map((record) => ({
			prompt: record.prompt,
			username: record.author,
		  }));
      

      return {
        records: generatedPrompts,
      };
```

Hopefully I can get this sorted out today!

**Comment by DiceBreakers** [comment, marked as answer]
I've got the kinks worked out! The client goes through a few steps to parse the data, I feel like a better programmer could probably clean this up somehow, but I'm at least happy to check it off the list for now and can worry about optimizing down the road when I have a better overall understanding.

I think my only remaining question is if it's safe to keep my View API rules blank for the user collection.

Hopefully this helps someone else down the line:

Client side code:

```
<script lang="ts">
	import '../app.postcss';
	import { catList } from './components/catList.svelte';
	import { Accordion, AccordionItem } from '@skeletonlabs/skeleton';
	import { onMount } from 'svelte';
	import { writable } from 'svelte/store';
	
	type CatItem = {
	  value: string;
	  label: string;
	  checked: boolean;
	};
	
	let selectedCategories = writable<CatItem[]>([]);
	let currentPrompt = "";
	let currentAuthor = "";
	let promptIndex = 0;
	let generatedPrompts: { prompt: string; author: string }[] = [];
	
	onMount(() => {
	  catList.subscribe((list: CatItem[]) => {
		selectedCategories.set(list);
	  });
	});
	
	async function generate(event: Event) {
  event.preventDefault();

  const formData = new FormData();

  selectedCategories.subscribe((list: CatItem[]) => {
    list.forEach(catItem => {
      if (catItem.checked) {
        formData.append('categories', catItem.value);
      }
    });
  });

  try {
    const response = await fetch('/play?/generate', {
      method: 'POST',
      body: formData,
    });

	if (response.ok) {
  const serverData = await response.json();
  const rawData = JSON.parse(serverData.data);

  // The prompt data is inside the array, so let's access it directly
  const promptData = JSON.parse(rawData[1]);

  generatedPrompts = promptData.map(obj => ({
    prompt: obj.prompt,
    author: obj.author,
  }));


      // Display the first prompt when generated
      displayNextPrompt();
    } else {
      console.error('Failed to generate prompts.');
    }
  } catch (error) {
    console.error('An error occurred:', error);
  }
}

	
function displayNextPrompt() {
  if (promptIndex < generatedPrompts.length) {
    const { prompt, author} = generatedPrompts[promptIndex];
    currentPrompt = prompt;
    currentAuthor = author; // Assuming you have a variable for displaying the author
    promptIndex++;
  } else {
    currentPrompt = "No more prompts"; // Optional message when there are no more prompts
  }
}
  </script>
```

Server-side:

```
import { error } from '@sveltejs/kit';

export const actions = {
  generate: async ({ request, locals }) => {
    const data = await request.formData();
    const selectedCategories = data.getAll('categories') ?? [];

    try {

        // Convert the category value to a string and then replace double quotes
        const categoriesString = selectedCategories.toString().replace(/,/g, '"||categories~"');

		let generatedPrompts = [];

		let records = await locals.pb.collection('prompts').getFullList({
      filter: `(categories~"${categoriesString}") && postPublic = true`,
      expand: 'author',
      fields: 'expand.author.username,prompt',
      sort: '-created',
			skipTotal: 1,
        });

		generatedPrompts = records.map((record) => ({
			prompt: record.prompt,
			author: record.expand.author.username,
		  }));
      

      return {
        records: JSON.stringify(generatedPrompts), // Convert to JSON string
      };
    } catch (err) {
      console.error('Error: ', err);
      throw error(err.status, "The robots didn't like something about that...");
    }
  },
};

```
